### PR TITLE
RecoTracker/ TkNavigation: fix clang warning -Woverloaded-virtual

### PR DESCRIPTION
**Pull request by gartung**
resubmit of #22765 and #22788 cherry-picked onto one branch.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoTracker/TkNavigation/plugins/CosmicNavigationSchool.cc [hunk 1]
```diff
@@ -44,6 +44,7 @@ class dso_hidden CosmicNavigationSchool : public SimpleNavigationSchool {
   //FakeDetLayer* theFakeDetLayer;
   void linkBarrelLayers( SymmetricLayerFinder& symFinder) override;
   //void linkForwardLayers( SymmetricLayerFinder& symFinder); 
+  void establishInverseRelations() override {};
```
Comment (slava77) [inline RecoTracker/TkNavigation/plugins/CosmicNavigationSchool.cc]: Please clarify why this change is applied.
If the intention is to disallow calls to this method from the base pointer, then better add an assert or an exception.
Otherwise, the original 
```using SimpleNavigationSchool::establishInverseRelations;``` seems to make more sense.

A further improvement can be to replace the implementation of ```void establishInverseRelations( SymmetricLayerFinder& symFinder )``` with 
```c++
void establishInverseRelations( SymmetricLayerFinder& symFinder ){
  SimpleNavigationSchool::establishInverseRelations();
  buildAdditionalForwardLinks(symFinder); 
}
```

File: RecoTracker/TkNavigation/plugins/SimpleNavigableLayer.h [hunk 2]
```diff
@@ -32,6 +32,13 @@ class dso_hidden SimpleNavigableLayer : public NavigableLayer {
 
   void setCheckCrossingSide(bool docheck) {theCheckCrossingSide = docheck;}
 
+  std::vector<const DetLayer*> 
+  compatibleLayers( NavigationDirection direction) const override = 0;
+
+  std::vector<const DetLayer*> 
+  compatibleLayers( const FreeTrajectoryState& fts, 
+		    PropagationDirection timeDirection) const override {int counter =0; return compatibleLayers(fts,timeDirection,counter);};
```
Comment (slava77) [inline RecoTracker/TkNavigation/plugins/SimpleNavigableLayer.h]: shouldn't this instead be ```return NavigableLayer::compatibleLayers(fts, timeDirection); ``` ?

Also, why was this a bad choice ```using NavigableLayer::compatibleLayers;``` ?

Comment (gartung) [reply RecoTracker/TkNavigation/plugins/SimpleNavigableLayer.h]: The classes derived from SimpleNavigableLayer did not see the function declaration from its base class NavigableLayer.

File: RecoTracker/TkNavigation/plugins/CosmicNavigationSchool.cc (hunk 1 above)
Comment (gartung) [reply RecoTracker/TkNavigation/plugins/CosmicNavigationSchool.cc]: This is another case where the derived classes do not see the function declarations for the base class SimpleNavigationSchool.

File: RecoTracker/TkNavigation/plugins/SimpleNavigableLayer.h (hunk 2 above)
Comment (gartung) [reply RecoTracker/TkNavigation/plugins/SimpleNavigableLayer.h]: Disregard previous statement.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/22790	76e5d14
- technical
- jenkins tests pass

**Comment by fabiocos** [comment]
+1
